### ISSUE TITLE
CUDA Error: (null) FasterTransformer/3rdparty/trt_fused_multihead_attention/fused_multihead_attention.h 232

### ISSUE THREAD

**Issue by DogeWatch** [NVIDIA/FasterTransformer#408]
when I call the function torch.classes.FasterTransformer.ParallelGptContextDecoderOp, here is the error message
<img width="1259" alt="image" src="https://user-images.githubusercontent.com/13670813/209674767-d69af705-108b-4b3e-bcf8-d043cf91b682.png">

environment is: cuda=11.0, openmpi=4.1.4, nccl=2.10.3-1, cudnn=8.4.1.50, torch=1.12.0

**Comment by byshiue** [comment]
Please follow the template of issue to provide the settings and reproduced steps.
Besides, please also try the main branch.

**Comment by DogeWatch** [comment]
> Please follow the template of issue to provide the settings and reproduced steps. Besides, please also try the main branch.

I've update my issue and use the main branch but still error

**Comment by byshiue** [comment]
It may be because you use virtual gpu and fail to load the kernel. 

You can try 
1. Find a environment with real GPU to test first. Or
2. Set the `is_fuse` to be false at https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/th_op/multi_gpu_gpt/ParallelGptOp.h#L292 to disable the kernels.

**Comment by DogeWatch** [comment]
> It may be because you use virtual gpu and fail to load the kernel.
> 
> You can try
> 
> 1. Find a environment with real GPU to test first. Or
> 2. Set the `is_fuse` to be false at https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/th_op/multi_gpu_gpt/ParallelGptOp.h#L292 to disable the kernels.

it works fine on read GPU, but still error when I set the `is_fuse` to be false and rebuild on virtual gpu

**Comment by byshiue** [comment]
Does the error affect the program? It happens during initializing the fused mha wrapper, but will not use it if the is_fuse is set to false. 
You can remove https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/layers/attention_layers/GptContextAttentionLayer.cc#L437 and https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/layers/attention_layers/GptContextAttentionLayer.cc#L470 to prevent the initialization.

**Comment by DogeWatch** [comment]
> Does the error affect the program? It happens during initializing the fused mha wrapper, but will not use it if the is_fuse is set to false. You can remove https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/layers/attention_layers/GptContextAttentionLayer.cc#L437 and https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/layers/attention_layers/GptContextAttentionLayer.cc#L470 to prevent the initialization.

It looks like solved the problem, but I found that the value with fp16 mode is not equal to the normal situation (real gpu and nvidia docker) with fp16 mode

**Comment by byshiue** [comment]
Do you disable the kernels for both cases? The kernels are never used in fp32 mode.

**Comment by DogeWatch** [comment]
> Do you disable the kernels for both cases? The kernels are never used in fp32 mode.

no, the virtual gpu side have disabled the fused attention kernel, and the real gpu side didn't do that

**Comment by byshiue** [comment]
Then they use different kernels and the different results are expected.

**Comment by DogeWatch** [comment]
> Then they use different kernels and the different results are expected.

Thank you for your answer, I found that the error I encountered may not be caused by vGPU, but caused by the environment dependencies I builded. Because I use my image also get this error on a real gpu, but using nvidia's docker image doesn't

nvidia's docker image I used is cuda 11.7, I don't know if cuda 11.0 works fine in your tests

**Comment by byshiue** [comment]
cuda 11.0 should work, too. You can try different version of docker image. More details are showed in https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html.
